src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -1161,10 +1164,13 @@ CallGenerator* CallGenerator::for_method_handle_inline(JVMState* jvms, ciMethod*
           ciType* t = signature->type_at(i);
           if (t->is_klass()) {
```

Comment:
What I suggested is to extend the guard with `!t->as_klass()->is_interface()` check.

Reply:
I am not convinced it is safe. At least it does not look safe for JDK 17. The "must" in the comment block above makes me wary about skipping the casting for interfaces...

```
        // In lambda forms we erase signature types to avoid resolving issues
        // involving class loaders.  When we optimize a method handle invoke
        // to a direct call we must cast the receiver and arguments to its
        // actual types.
```

Reply:
Not sure what safety concerns you have in mind w.r.t interfaces. On bytecode level interface treatment is lax (e.g., verifier ignores interface type info; see JVMS-4.10.1.2 for details).

Reply:
I am mostly concerned about pushing a "less than obvious and wider than definitely required" fix to JDK 17. Filtering out by "empty join type" is quite probably too wide. Filtering out the interface arguments is much less obvious. Yes, I get that interface arguments in signatures might be filtered wholesale, but I would rather wait for more testing to see if that actually works.

Reply:
Frankly speaking, I wouldn't expect too much from testing considering the problem went unnoticed for 2,5 years.

But, as an additional data point, I'd like to reiterate that the fix for JDK-8268371 applied a similar strategy in a much more common scenario - inlining of interface methods - and so far there were no problems spotted.